### PR TITLE
Speeding up `perfect_power`

### PR DESCRIPTION
**Pull request by haru-44**
If n is small, the answer is obtained by floating-point arithmetic instead of using `integer_nthroot`.

#### References to other Issues or PRs

#25317

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/ntheory/factor_.py [hunk 1]
```diff
@@ -495,10 +495,12 @@ def perfect_power(n, candidates=None, big=True, factor=True):
         # no unique exponent for 0, 1
         # 2 and 3 have exponents of 1
         return False
-    logn = math.log(n, 2)
-    max_possible = int(logn) + 2  # only check values less than this
-    not_square = n % 10 in [2, 3, 7, 8]  # squares cannot end in 2, 3, 7, 8
-    min_possible = 2 + not_square
+    logn = math.log2(n)
+    max_possible = int(logn) + 1  # only check values less than this
+    if 0x213 & (1 << (n & 15)): # simple square number check
```
Comment (smichr) [inline sympy/ntheory/factor_.py:500]: Could you give a reference to this, please? I checked several mathexchange pages and did not find this.

(saw this, though: https://math.stackexchange.com/a/712818/122921)

Comment (smichr) [reply sympy/ntheory/factor_.py:500]: See also http://cr.yp.to/lineartime/powers2-20060914-ams.pdf

**Comment by smichr** [comment]
When I time `pp4=[i for i in range(10**5,10**6) if perfect_power(i)]` for this version and the one in master, this version takes about 25% longer.

**Comment by haru-44** [comment]
> When I time `pp4=[i for i in range(10**5,10**6) if perfect_power(i)]` for this version and the one in master, this version takes about 25% longer.

Oh, I see. The current version is faster when a small factor can be expected to be present. For example, the PR version should be faster for `pp4=[i for i in range(10**5+1,10**6, 2) if perfect_power(i)]`.

Removing lines 520-533 would make it faster for any input, but it would only be a small improvement. I would like to delete these lines.

File: sympy/ntheory/factor_.py (hunk 1 above)
Comment (haru-44) [reply sympy/ntheory/factor_.py:500]: I would like to read the paper you presented later.

The square number determination was based on the following code.

https://github.com/sympy/sympy/blob/d59d9bd65c552ddc763071def5893f231c1f8294/sympy/external/gmpy.py#L143-L155

There is a problem of which number to give to `magic`, but here I used 16. The current version determines that `4/10` is not a square number, whereas PR's version determines that `12/16` is not a square number.

Comment (haru-44) [reply sympy/ntheory/factor_.py:500]: The paper you have presented may be approximately fast, but it does not seem valid in practical numbers.
In step 1, we are looking for $r_q$. If we can find this, we can check whether $r_q^q$ is $n$. Indeed, for `n=7**(3**k)`, we have to perform a recursion of depth `k`. However, the gain from improving such a case seems marginal.

**Comment by sylee957** [comment]
I think that benchmark assertion should be to given to test these sort of performance improvements which are not trivial.

**Pull request closed without merging**